### PR TITLE
ci: Implement Replay Checker

### LINKED ISSUE #1201: Check replays in CI
Now that we have replay checking in headless mode and via commandline, we should use that in CI for automatic compatibility checking.

I have curated a list of replays. It started with the `golden replay #1` and I added a replay each time a mismatch was discovered and where the mismatch was not detected by one of the existing list of replays:
[replay_list_min.zip](https://github.com/user-attachments/files/20931219/replay_list_min.zip)

It currently contains 9 replays and only requires official maps. I also tried to select replays that can be simulated quickly. I think it would make sense to use these replays in CI. We can still check more replays offline and expand that set of replays if more mismatches are discovered.

In CI, those replays first need to be unpacked into the replay folder. The game binary then needs to be called like this:
```
generalszh.exe -jobs 2 -headless -replay replay_list_min/*.rep
```
The return value needs to be evaluated (1 means mismatch or crash in one or more replays).

I'd also like to see another build configuration added where DEBUG_LOGGING and DEBUG_CRASHING are enabled (but not RTS_DEBUG or RTS_INTERNAL). That configuration should also be checked for compatibility.

### PR DESCRIPTION
**Pull request by helmutbuhler**
Fixes #1201
Initial attempt here: #1239

This PR adds automatic checking of replays for compatibility to CI.
It also adds an additional configuration `vc6-releaselog` to CMakePresets.json. It compiles the game with optimizations, logging and crashing enabled. That configuration is still compatible and also checked in the CI.

It also adds two more headless checks:
- One disables a dialog that appears when the music cannot be loaded (not included in the gamedata files CI uses)
- The other is needed because `MessageBoxWrapper` since a recent change no longer checks whether there is a window in the current thread.

The CI internally downloads some gamedata files which are needed to simulate the replay. These gamedata files are trimmed of files which are not needed in headless mode (audio, textures, GUI). They are also encrypted so only the CI runner can access them.

The replays and maps the CI uses to check are also added here to the repository.

The new replaycheck job happens after the VC6 build is complete. Luckily, the VC6 build is currently much faster than the VS22 build and as such the replaycheck and the VS22 build can run in parallel. This PR therefore doesn't increase the runtime of CI much (about a minute).

Before this can be merged, these things still need to happen:
- ~~ArcticDolphin needs to upload the file to s3~~
- ~~Issue #1365 needs to be fixed~~

### CODE REVIEW

File: .github/workflows/check-replays.yml [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+          if (-not $env:AWS_ACCESS_KEY_ID -or -not $env:AWS_SECRET_ACCESS_KEY -or -not $env:AWS_ENDPOINT_URL) {
+              $ok1 = [bool]$env:AWS_ACCESS_KEY_ID
+              $ok2 = [bool]$env:AWS_SECRET_ACCESS_KEY
+              $ok3 = [bool]$env:AWS_ENDPOINT_URL
+              Write-Host "One or more required secrets are not set or are empty. R2_ACCESS_KEY_ID: $ok1, R2_SECRET_ACCESS_KEY: $ok2, R2_ENDPOINT_URL: $ok3"
+              exit 1
+          }
+          Write-Host "Downloading GameData" -ForegroundColor Cyan
+          aws s3 cp s3://github-ci/gen_gamedata_trimmed.7z gen_gamedata_trimmed.7z --endpoint-url $env:AWS_ENDPOINT_URL
```
Comment (xezon) [inline .github/workflows/check-replays.yml]: How big is this data? Will this download anew on every run or is this kept cached on the ci runner?

File: .github/workflows/check-replays.yml [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+      - name: Download Game Data from Cloudflare R2
+        if: ${{ steps.cache-gamedata.outputs.cache-hit != 'true' }}
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.R2_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.R2_SECRET_ACCESS_KEY }}
+          AWS_ENDPOINT_URL: ${{ secrets.R2_ENDPOINT_URL }}
+          EXPECTED_HASH: "3BE86F84B1E771BEA01BD2DBA39442683381AE4401716E9AE9D4AD95C8F45372"
+        shell: pwsh
+        run: |
+          # Download trimmed gamedata of both ZH and Generals. This data cannot be used for playing because it's
```
Comment (xezon) [inline .github/workflows/check-replays.yml]: "Generals 1.08 and Generals Zero Hour 1.04"

I suggest to always name Generals first for consistency.

File: .github/workflows/check-replays.yml (hunk 1 above)
Comment (xezon) [inline .github/workflows/check-replays.yml]: perhaps name this file

"generals108_zh104_gamedata_trimmed.7z"

to make clear that this file contains not just Generals.

File: GeneralsMD/Test/Replays/366648.rep
Comment (xezon) [inline GeneralsMD/Test/Replays/366648.rep:1]: Does any of these replays contain uncensored IP addresses?

Note that Replay headers contain IP addresses. We probably want to eliminate these before merging them into the repository.

The IP Addresses in GenTool Replays are zeroed in the lower 12 bits.

File: GeneralsMD/Test/README.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Test folder
+
+This folder contains replays and the required maps that are tested in CI to ensure that the game is retail compatible.
+
+You can also test with these replays locally:
+- Copy the replays into a subfolder in your `*userpath*/Documents/Command and Conquer Generals Zero Hour Data/Replays` folder.
+- Copy the maps into `*userpath*/Documents/Command and Conquer Generals Zero Hour Data/Maps`
```
Comment (xezon) [inline GeneralsMD/Test/README.md]: `%USERPROFILE%\Documents\...`

File: Generals/Code/GameEngine/Source/Common/System/Debug.cpp [hunk 4]
```diff
@@ -148,11 +148,14 @@ static void doStackDump();
...
 #ifdef DEBUG_CRASHING
-	return !DX8Wrapper_IsWindowed || (TheGlobalData&&TheGlobalData->m_debugIgnoreAsserts);
-#else
-	return !DX8Wrapper_IsWindowed;
+	if (TheGlobalData && TheGlobalData->m_debugIgnoreAsserts)
+		ignore = true;
 #endif
+	if (TheGlobalData && TheGlobalData->m_headless)
+		ignore = true;
+	return ignore;
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/Common/System/Debug.cpp]: I think this can be further simplified:

```cpp
inline Bool ignoringAsserts()
{
	if (!DX8Wrapper_IsWindowed)
		return true;
	if (TheGlobalData && TheGlobalData->m_headless)
		return true;
#ifdef DEBUG_CRASHING
	if (TheGlobalData && TheGlobalData->m_debugIgnoreAsserts)
		return true;
#endif

	return false;
}

**Review by xezon — COMMENTED** [review]
Very nice.

@OmniBlade Are you ok with 3 MB of game user maps and replays in code repo? Or should this be stored a different way?

File: GeneralsMD/Test/Replays/366648.rep
Comment (barefootlogician) [reply GeneralsMD/Test/Replays/366648.rep:1]: WHAT! THAT'S INSANE

File: GeneralsMD/Code/GameEngine/Source/Common/System/Debug.cpp
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Source/Common/System/Debug.cpp:1]: For cleaner separation of changes, I think it would be good to have 2 commits go into main branch:

1. The C++ code changes
2. yml edits and binary files additions

File: .github/workflows/check-replays.yml (hunk 1 above)
Comment (helmutbuhler) [reply .github/workflows/check-replays.yml]: 105 MB. It's only downloaded once when the PR will be merged and after that loaded from the cache.

Comment (helmutbuhler) [reply .github/workflows/check-replays.yml]: Too late :) Already uploaded.

File: GeneralsMD/Test/Replays/366648.rep
Comment (helmutbuhler) [reply GeneralsMD/Test/Replays/366648.rep:1]: I agree that it's insane...
I haven't checked whether they have valid ip addresses, but all replays (except for the Golden one) are from the gentool server, so they should be "censored".
Do you have a tool to reset the ip address of a replay? If so you can push a commit with new replays. Just make sure the ip is set to something !=0 because there is some code that checks for that. (Maybe set it to 1 or 42).

Comment (barefootlogician) [reply GeneralsMD/Test/Replays/366648.rep:1]: Maybe find and replace with regex would suffice for removing the ip addresses

**Comment by helmutbuhler** [comment]
@tintinhamans Thank you for uploading the datafile to S3 and the help!

@xezon and @Caball009 Thank you for fixing #1365!

I just pulled main and also implemented all suggestions (except for the ip stuff). The CI works now! I also extracted the code changes as suggested into a new PR #1378.

For security reasons, the secrets are not visible for PRs, that's why the CI still seems to fail here. But once it's merged (xezon will have to override the CI check), it should work. After the merge it will also work for all other PRs because the gamedata will be in the cache and not require the secrets anymore.

If you want to check what the CI run will look like, you can see it here: https://github.com/helmutbuhler/CnC_Generals_Zero_Hour/pull/1
(I also tested it there with mismatching builds, timeouts and all those edge cases)

File: GeneralsMD/Test/Replays/366648.rep
Comment (helmutbuhler) [reply GeneralsMD/Test/Replays/366648.rep:1]: It's binary

Comment (barefootlogician) [reply GeneralsMD/Test/Replays/366648.rep:1]: > It's binary

Try hex editor?

Comment (helmutbuhler) [reply GeneralsMD/Test/Replays/366648.rep:1]: Maybe you wanna try it? I can give you permission to push to this branch.

Comment (JoKeRZH429) [reply GeneralsMD/Test/Replays/366648.rep:1]: I checked. Golden Replay is not censored.

File: .github/workflows/check-replays.yml (hunk 1 above)
Comment (xezon) [reply .github/workflows/check-replays.yml]: Cannot be renamed ever?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -143,3 +163,19 @@ jobs:
       tools: ${{ matrix.tools }}
       extras: ${{ matrix.extras }}
     secrets: inherit
+
+  replaycheck-generalsmd:
+    name: Replay Check MD${{ matrix.preset && '' }}
```
Comment (xezon) [inline .github/workflows/ci.yml]: "Replay Check GeneralsMD"

File: .github/workflows/check-replays.yml (hunk 1 above)
Comment (helmutbuhler) [reply .github/workflows/check-replays.yml]: @tintinhamans can rename it.

Comment (xezon) [reply .github/workflows/check-replays.yml]: Also, perhaps it makes sense to split this into 2 files:

"generals108_gamedata_trimmed.7z"
"zerohour104_gamedata_trimmed.7z"

This then gives the freedom to combine this with another zero hour data version. Just a bit more flexible.

**Comment by OmniBlade** [comment]
> Very nice.
> 
> @OmniBlade Are you ok with 3 MB of game user maps and replays in code repo? Or should this be stored a different way?

Hmm, depends on if its static data, we already have a few tiny blobs of binary data with things like icons present and abit of data for testing isn't out of the question, but perhaps doing it the same as the game data and pulling it in would be a better idea. Especially for the replay data as that will need to be regenerated whenever we create incompatible changes to the code base.

**Comment by xezon** [comment]
We could add the data here:

https://github.com/TheSuperHackers/GeneralsGameData/

It is meant to be a game data dump.

**Comment by barefootlogician** [comment]
Why not just store it in something like Google drive

**Comment by xezon** [comment]
It is already bad enough to have things in Amazon cloud. I suggest to not spread data around and keep most things inside the github organization when reasonable. It makes it easier to access and maintain.

I have no access to the Amazon cloud.

**Comment by helmutbuhler** [comment]
> We could add the data here:
> 
> https://github.com/TheSuperHackers/GeneralsGameData/
> 
> It is meant to be a game data dump.

I prefer to have it in git. And I already mentioned in discord:
> Stored in plain git without lfs. lfs doesn't work in github forks but that's fine because it's less than 3MB.
That way it will be easy for other developers to locally test branches against the replays, we will see in git history for what reasons we included what replays and it all works without cloning other repos.
Any objections or suggestions?

You agreed and noone else said anything.

I already knew, whereever I stored those files, someone would come along and suggest to store them somewhere else. That's why I asked beforehand to avoid unnecessary work.

**Comment by xezon** [comment]
GeneralsGameData is also git. Just in a separate repo. The advantage is that we can just keep adding more and more data there without any limitations.

**Comment by OmniBlade** [comment]
If its a another git repo it can potentially be pulled in as a submodule or fetchcontent. You could also potentially abuse the github attachments to create a URL to it that is associated with the repo but not included in the git repo.

File: .github/workflows/check-replays.yml (hunk 1 above)
Comment (helmutbuhler) [reply .github/workflows/check-replays.yml]: If @tintinhamans does that, I'll update the code.

**Comment by helmutbuhler** [comment]
> GeneralsGameData is also git. Just in a separate repo. The advantage is that we can just keep adding more and more data there without any limitations.

I know it's also git. Both approaches have pros and cons.

> If its a another git repo it can potentially be pulled in as a submodule or fetchcontent. You could also potentially abuse the github attachments to create a URL to it that is associated with the repo but not included in the git repo.

github attachments are deleted after some time. But I guess it would work as a submodule.
Anyone wants to set that up?

**Comment by xezon** [comment]
I can help with that. It will be also a good learning excercise because I want to revisit and consider submodules for the Patch repository. (This code repo likely will be a submodule)

**Comment by xezon** [comment]
@helmutbuhler Check TheSuperHackers/GeneralsReplays#1

**Comment by helmutbuhler** [comment]
> @helmutbuhler Check [TheSuperHackers/GeneralsReplays#1](https://github.com/TheSuperHackers/GeneralsReplays/pull/1)

I left a comment there.

**Comment by xezon** [comment]
"GeneralsReplays" submodule was added.

File: .github/workflows/check-replays.yml (hunk 1 above)
Comment (xezon) [reply .github/workflows/check-replays.yml]: I would do as suggested above. Is flexible and identifiable.

**Comment by helmutbuhler** [comment]
Ok, I updated it again. The CI now uses the replays and maps here: https://github.com/TheSuperHackers/GeneralsReplays
The gamedata in S3 has also been renamed and split into two archives.

**Comment by xezon** [comment]
> One or more required secrets are not set or are empty. R2_ACCESS_KEY_ID: False, R2_SECRET_ACCESS_KEY: False, R2_ENDPOINT_URL: False

@tintinhamans Can you supply these things whereever they need to go?

**Comment by helmutbuhler** [comment]
> > One or more required secrets are not set or are empty. R2_ACCESS_KEY_ID: False, R2_SECRET_ACCESS_KEY: False, R2_ENDPOINT_URL: False
> 
> @tintinhamans Can you supply these things whereever they need to go?

They are already in this repository. It's just that they are hidden for pending PRs for security reasons. That's why it's failing currently. Once this PR is merged, all CI Jobs will work. Also all Pending PRs will work from then on, because the data will be in the cache.
If you want to inspect how the run will look, you can look here: https://github.com/helmutbuhler/CnC_Generals_Zero_Hour/pull/1

File: .github/workflows/check-replays.yml [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+          Copy-Item -Path $source\* -Destination $destination -Recurse -Force
+
+      - name: Set Generals InstallPath in Registry
+        shell: pwsh
+        run: |
+          # Zero Hour loads some Generals files and needs this registry key to find the
+          # Generals data files.
+
+          $regPath = "HKCU:\SOFTWARE\Electronic Arts\EA Games\Generals"
+          $installPath = "C:\GameData\Generals\"
```
Comment (xezon) [inline .github/workflows/check-replays.yml]: This script maintains multiple variables with the same strings:

1. `path: C:\GameData`
2. `-oC:\GameData\Generals`
3. `$installPath = "C:\GameData\Generals\"`
4. `$source = "C:\GameData\${{ inputs.game }}"`

I think it would be good to share the names.

So for example

```
$gamePath = "C:\GameData"
$generalsPath = "${{gamePath}}\Generals"
$generalsMDPath = "${{gamePath}}\GeneralsMD"
```

Then it will be easier to tweak locations without breaking things.

File: .github/workflows/check-replays.yml [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+          # BINKW32.DLL
+          # English.big
+          # INI.big
+          # Maps.big
+          # mss32.dll
+          # W3D.big
+          # Data\Scripts\MultiplayerScripts.scb
+          # Data\Scripts\SkirmishScripts.scb
+
+          Write-Host "Downloading GameData Generals" -ForegroundColor Cyan
```
Comment (xezon) [inline .github/workflows/check-replays.yml]: Maybe improve title as "Downloading Game Data for Generals"

File: .github/workflows/check-replays.yml [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        run: |
+          # These files are expected in the user dir, so we move them here.
+
+          $source = "${{ inputs.gamedata }}\Replays"
+          $destination = "$env:USERPROFILE\Documents\Command and Conquer Generals Zero Hour Data\Replays"
+          Write-Host "Move replays to $destination"
+          New-Item -ItemType Directory -Path $destination -Force | Out-Null
+          Move-Item -Path "$source\*" -Destination $destination -Force
+
+          $source = "${{ inputs.gamedata }}\Maps"
```
Comment (xezon) [inline .github/workflows/check-replays.yml]: The naming of `inputs.gamedata` is ambiguous with GameData, used for the static game data. Maybe name it `inputs.userdata`.

File: .github/workflows/check-replays.yml [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+  workflow_call:
+    inputs:
+      game:
+        required: true
+        type: string
+        description: "Game to check (only GeneralsMD for now)"
+      gamedata:
+        required: true
+        type: string
+        description: "path to folder with replays and maps"
```
Comment (xezon) [inline .github/workflows/check-replays.yml]: "Path to folder with replays and maps"

**Review by xezon — COMMENTED** [review]
Some minor comments

**Comment by helmutbuhler** [comment]
Implemented all suggestions

File: .github/workflows/check-replays.yml [hunk 10]
```diff
@@ -59,6 +59,10 @@ jobs:
           # missing textures, audio and gui files. But it's enough for replay checking.
           # It's also encrypted because it's not allowed to distribute these files.
 
+          $gamePath = "C:\GameData"
```
Comment (xezon) [inline .github/workflows/check-replays.yml]: Can you try move that into the `env:` so that all steps can use it? Then you can remove all the duplicates.

Put the `env:` before the `steps:`

```
    env:
      GAME_PATH: C:\GameData
      GENERALS_PATH ${{ env.GAME_PATH }}\Generals
      GENERALSMD_PATH ${{ env.GAME_PATH }}\GeneralsMD
```

Comment (helmutbuhler) [reply .github/workflows/check-replays.yml]: I considered that, but it doesn't seem to work. env: works only in steps.

Comment (helmutbuhler) [reply .github/workflows/check-replays.yml]: But you can try to make it work if you want to. You can push things here and see in my github if it works.

Comment (xezon) [reply .github/workflows/check-replays.yml]: Hmm. Chat GPT says:

---

It does work—**but only in the right places**. `env:` is valid at three levels:

1. **Workflow level** (applies to all jobs):

   ```yaml
   name: Demo
   on: workflow_dispatch
   env:
     GLOBAL: foo

   jobs:
     build:
       runs-on: ubuntu-latest
       steps:
         - run: echo "$GLOBAL"
   ```

2. **Job level** (applies to all steps in that job):

   ```yaml
   jobs:
     build:
       runs-on: ubuntu-latest
       env:
         BASE: /some/path
         ALIAS: ${{ env.BASE }}/subdir
       steps:
         - run: echo "$ALIAS"
   ```

3. **Step level** (only that step):

   ```yaml
   - name: One-off
     env:
       LOCAL: bar
     run: echo "$LOCAL"
   ```

Common pitfalls:

* **Wrong indentation/place**: `env:` must be a sibling of `runs-on` for jobs, or of `run`/`uses` for a step. If you put `env:` outside of a job/step in the wrong spot it won’t take effect.
* **Shell-style interpolation in YAML-level `env:`** doesn’t work; use expressions: `ALIAS: ${{ env.BASE }}/subdir`, not `ALIAS: $BASE/subdir`.
* **Referencing an env var before it’s defined**: at the same level you can reference earlier ones (e.g., `ALIAS` uses `BASE`), but you can’t do something like `BASE: ${{ env.ALIAS }}` if `ALIAS` is defined later (ordering matters conceptually, though GitHub evaluates the block).

If you share what your failing snippet looks like, I can point out the exact issue.

Comment (helmutbuhler) [reply .github/workflows/check-replays.yml]: I took some codechanges from my ChatGPT, let's see if it works.

Comment (xezon) [reply .github/workflows/check-replays.yml]: Hopefully. It looks good to the eye :)

Comment (helmutbuhler) [reply .github/workflows/check-replays.yml]: Works! My ChatGPT from earlier gave some wrong code.

**Review by xezon — APPROVED** [review]

**Comment by xezon** [comment]
GitHub says "This branch cannot be rebased due to conflicts". Cannot merge?

**Comment by helmutbuhler** [comment]
Just pushed merge of main into this PR. Maybe that helps.

**Comment by Mauller** [comment]
This got merged with a bad commit title
